File: packages/protocol-client/rust/lib/stream.rs
//! Owned subscriptions, split senders, and one consuming receiver per ID.

use std::sync::{Arc, atomic::Ordering};

use tokio::sync::mpsc;

use crate::router::{Lease, QueuedFrame};
use crate::{Client, ClientResult, Delivery, IntoOutboundMessage, Message, Protocol, RawFrame};

//--------------------------------------------------------------------------------------------------
// Types
//--------------------------------------------------------------------------------------------------

/// Opaque frame stream whose lifetime retains its connection and ID lease.
pub struct RawStream<P: Protocol> {
    sender: RawStreamSender<P>,
    pub(crate) receiver: RawStreamReceiver<P>,
}

/// Cloneable send permission bound to one specific lease, not merely its number.
pub struct RawStreamSender<P: Protocol> {
    client: Client<P>,
    lease: Arc<Lease>,
}

/// Sole consuming raw receiver. Dropping it disables sends and starts draining.
pub struct RawStreamReceiver<P: Protocol> {
    client: Client<P>,
    pub(crate) lease: Arc<Lease>,
    receiver: mpsc::Receiver<QueuedFrame>,
    done: bool,
}

/// Decoded message stream over the same opaque router and lease.
pub struct Stream<P: Protocol> {
    raw: RawStream<P>,
}

/// Cloneable native-message sender bound to an owned stream lease.
pub struct StreamSender<P: Protocol> {
    raw: RawStreamSender<P>,
}

/// Sole decoded receiver; no competing reader is created by conversion/split.
pub struct StreamReceiver<P: Protocol> {
    raw: RawStreamReceiver<P>,
}

//--------------------------------------------------------------------------------------------------
// Methods
//--------------------------------------------------------------------------------------------------

impl<P: Protocol> RawStream<P> {
    pub(crate) fn new(
        client: Client<P>,
        lease: Arc<Lease>,
        receiver: mpsc::Receiver<QueuedFrame>,
    ) -> Self {
        Self {
            sender: RawStreamSender {
                client: client.clone(),
                lease: Arc::clone(&lease),
            },
            receiver: RawStreamReceiver {
                client,
                lease,
                receiver,
                done: false,
            },
        }
    }

    /// Current stream's correlation ID.
    pub fn id(&self) -> u32 {
        self.sender.id()
    }

    /// Send a follow-up opaque frame; local close never sends a protocol signal.
    pub async fn send(&self, flags: u8, body: &[u8]) -> ClientResult<()> {
        self.sender.send(flags, body).await
    }

    /// Receive each frame, terminal included, then clean exhaustion or an error.
    pub async fn recv(&mut self) -> ClientResult<Option<RawFrame>> {
        self.receiver.recv().await
    }

    /// Abandon local consumption and retain bounded drain state until terminal.
    pub fn close(&mut self) {
        self.receiver.close();
    }

    /// Move ownership into a cloneable sender and one consuming receiver.
    pub fn into_parts(self) -> (RawStreamSender<P>, RawStreamReceiver<P>) {
        (self.sender, self.receiver)
    }
}

impl<P: Protocol> RawStreamSender<P> {
    /// Correlation ID associated with this specific lease.
    pub fn id(&self) -> u32 {
        self.lease.id
    }

    /// Send while this lease remains live; a reused numeric ID gives no authority.
    pub async fn send(&self, flags: u8, body: &[u8]) -> ClientResult<()> {
        self.client.send_raw_owned(&self.lease, flags, body).await
    }
}

impl<P: Protocol> RawStreamReceiver<P> {
    /// Correlation ID whose frames this receiver alone consumes.
    pub fn id(&self) -> u32 {
        self.lease.id
    }

    /// Receive the next frame. EOF before terminal is an error exactly once.
    pub async fn recv(&mut self) -> ClientResult<Option<RawFrame>> {
        if self.done {
            return Ok(None);
        }
        match self.receiver.recv().await {
            Some(queued) => {
                if queued.frame.flags & microsandbox_protocol::message::FLAG_TERMINAL != 0 {
                    self.done = true;
                }
                Ok(Some(queued.frame))
            }
            None => {
                self.done = true;
                if self.lease.terminal.load(Ordering::Acquire) {
                    Ok(None)
                } else {
                    Err(self
                        .client
                        .inner
                        .state
                        .error()
                        .with_delivery(self.lease.delivery()))
                }
            }
        }
    }

    /// Stop local consumption and disable sends without canceling remote work.
    pub fn close(&mut self) {
        self.client.inner.state.abandon(&self.lease);
        self.receiver.close();
        while self.receiver.try_recv().is_ok() {}
        self.done = true;
    }
}

impl<P: Protocol> Stream<P> {
    pub(crate) fn from_raw(raw: RawStream<P>) -> Self {
        Self { raw }
    }

    /// Correlation ID of the owned stream.
    pub fn id(&self) -> u32 {
        self.raw.id()
    }

    /// Send a native or already-encoded payload using the selected codec.
    pub async fn send<M: IntoOutboundMessage<P>>(&self, message: M) -> ClientResult<()> {
        self.raw
            .sender
            .client
            .send_owned(&self.raw.sender.lease, message)
            .await
    }

    /// Receive an inspectable message, including unknown names and peer errors.
    pub async fn recv(&mut self) -> ClientResult<Option<Message>> {
        self.raw
            .recv()
            .await?
            .map(|frame| {
                self.raw
                    .sender
                    .client
                    .inner
                    .codec
                    .decode(frame)
                    .map_err(|error| error.with_delivery(Delivery::Unknown))
            })
            .transpose()
    }

    /// Abandon consumption; no domain-specific cancellation or EOF is sent.
    pub fn close(&mut self) {
        self.raw.close();
    }

    /// Split into owned handles while preserving one receiver and the ID lease.
    pub fn into_parts(self) -> (StreamSender<P>, StreamReceiver<P>) {
        let (sender, receiver) = self.raw.into_parts();
        (
            StreamSender { raw: sender },
            StreamReceiver { raw: receiver },
        )
    }
}

impl<P: Protocol> StreamSender<P> {
    /// Correlation ID of this lease.
    pub fn id(&self) -> u32 {
        self.raw.id()
    }

    /// Send on this lease, rejecting stale senders even after numeric ID reuse.
    pub async fn send<M: IntoOutboundMessage<P>>(&self, message: M) -> ClientResult<()> {
        self.raw.client.send_owned(&self.raw.lease, message).await
    }
}

impl<P: Protocol> StreamReceiver<P> {
    /// Correlation ID consumed by this receiver.
    pub fn id(&self) -> u32 {
        self.raw.id()
    }

    /// Receive messages on the sole raw receiver, decoding only on demand.
    pub async fn recv(&mut self) -> ClientResult<Option<Message>> {
        self.raw
            .recv()
            .await?
            .map(|frame| {
                self.raw
                    .client
                    .inner
                    .codec
                    .decode(frame)
                    .map_err(|error| error.with_delivery(Delivery::Unknown))
            })
            .transpose()
    }

    /// Disable further sends and drain until terminal completion.
    pub fn close(&mut self) {
        self.raw.close();
    }
}

//--------------------------------------------------------------------------------------------------
// Trait Implementations
//--------------------------------------------------------------------------------------------------

impl<P: Protocol> Clone for RawStreamSender<P> {
    fn clone(&self) -> Self {
        Self {
            client: self.client.clone(),
            lease: Arc::clone(&self.lease),
        }
    }
}

impl<P: Protocol> Clone for StreamSender<P> {
    fn clone(&self) -> Self {
        Self {
            raw: self.raw.clone(),
        }
    }
}

impl<P: Protocol> Drop for RawStreamReceiver<P> {
    fn drop(&mut self) {
        self.client.inner.state.abandon(&self.lease);
    }
}

File: packages/protocol-client/typescript/src/stream.ts
import { Core, type Lease } from "./core.js";
import { ClientError, clientError } from "./error.js";
import type { InboundFrame, OutboundMessage } from "./message.js";
import type { Protocol, RequestOptions } from "./protocol.js";
import { Attempt } from "./timing.js";
import { FLAG_TERMINAL, type RawFrame } from "./wire.js";

// Native wrappers need codec access, but public stream handles must not expose
// the mutable router, ownership records, or another stream's permissions.
const codecs = new WeakMap<object, {
  prepare(message: OutboundMessage): { flags: number; body: Uint8Array };
  decode(frame: RawFrame): InboundFrame;
}>();
function rememberCodec<P extends Protocol>(owner: object, core: Core<P>): void {
  codecs.set(owner, {
    prepare: message => core.prepare(message),
    decode: frame => core.established.codec.decode(frame),
  });
}

/** Cloneable send permission bound to a lease object, not just its numeric ID. */
export class RawStreamSender<P extends Protocol> {
  private readonly release: () => void;
  private closed = false;
  constructor(private readonly core: Core<P>, private readonly lease: Lease) {
    this.release = core.retain(this);
    rememberCodec(this, core);
  }
  get id(): number { return this.lease.id; }
  clone(): RawStreamSender<P> { if (this.closed) throw new ClientError("stream_closed"); return new RawStreamSender(this.core, this.lease); }
  async send(flags: number, body: Uint8Array, options: RequestOptions = {}): Promise<void> {
    if (this.closed) throw new ClientError("stream_closed");
    const attempt = new Attempt(options.requestTimeoutMs, options.signal);
    try { await this.core.writeFrame(this.lease, flags, body, attempt.signal); }
    finally { attempt.close(); }
  }
  close(): void { if (!this.closed) { this.closed = true; this.release(); } }
}

/** One consuming receiver. Abandoning it retains draining IDs until terminal. */
export class RawStreamReceiver<P extends Protocol> implements AsyncIterable<RawFrame> {
  private readonly release: () => void;
  private done = false;
  private receiving = false;
  constructor(private readonly core: Core<P>, private readonly lease: Lease) {
    this.release = core.retain(this, lease);
    rememberCodec(this, core);
  }
  get id(): number { return this.lease.id; }
  async next(options: RequestOptions | number = {}): Promise<RawFrame | null> {
    if (this.done) return null;
    if (this.receiving) throw new ClientError("receiving");
    const settings = typeof options === "number" ? { requestTimeoutMs: options } : options;
    const attempt = new Attempt(settings.requestTimeoutMs, settings.signal);
    this.receiving = true;
    try {
      const frame = await this.lease.queue.next(attempt.signal);
      if (frame === null || (frame.flags & FLAG_TERMINAL) !== 0) { this.done = true; this.release(); }
      return frame;
    } catch (error) {
      const failure = clientError(error).withDelivery(this.core.delivery(this.lease));
      if (failure.code !== "timeout" && failure.code !== "cancelled") { this.done = true; this.release(); }
      throw failure;
    } finally { this.receiving = false; attempt.close(); }
  }
  close(): void { this.core.abandon(this.lease); this.done = true; this.release(); }
  async *[Symbol.asyncIterator](): AsyncIterator<RawFrame> {
    try { for (;;) { const frame = await this.next(); if (!frame) return; yield frame; } }
    finally { this.close(); }
  }
}

/** Opaque stream. Splitting consumes this object's authority to send or receive. */
export class RawStream<P extends Protocol> implements AsyncIterable<RawFrame> {
  private parts?: { sender: RawStreamSender<P>; receiver: RawStreamReceiver<P> };
  constructor(core: Core<P>, lease: Lease) {
    this.parts = { sender: new RawStreamSender(core, lease), receiver: new RawStreamReceiver(core, lease) };
  }
  get id(): number { return this.owned().sender.id; }
  send(flags: number, body: Uint8Array, options?: RequestOptions): Promise<void> { return this.owned().sender.send(flags, body, options); }
  next(options?: RequestOptions | number): Promise<RawFrame | null> { return this.owned().receiver.next(options); }
  close(): void { if (this.parts) { this.parts.receiver.close(); this.parts.sender.close(); } }
  split(): { sender: RawStreamSender<P>; receiver: RawStreamReceiver<P> } {
    const parts = this.owned(); this.parts = undefined; return parts;
  }
  private owned() { if (!this.parts) throw new ClientError("split"); return this.parts; }
  async *[Symbol.asyncIterator](): AsyncIterator<RawFrame> {
    try { for (;;) { const frame = await this.next(); if (!frame) return; yield frame; } }
    finally { this.close(); }
  }
}

/** Native/encoded-payload sender using the selected protocol codec and gates. */
export class StreamSender<P extends Protocol> {
  constructor(private readonly raw: RawStreamSender<P>) {}
  get id(): number { return this.raw.id; }
  clone(): StreamSender<P> { return new StreamSender(this.raw.clone()); }
  async send(message: OutboundMessage, options?: RequestOptions): Promise<void> {
    const outbound = codecs.get(this.raw)!.prepare(message);
    await this.raw.send(outbound.flags, outbound.body, options);
  }
  close(): void { this.raw.close(); }
}

/** Native message receiver that decodes only when the consumer asks for a frame. */
export class StreamReceiver<P extends Protocol> implements AsyncIterable<InboundFrame> {
  constructor(private readonly raw: RawStreamReceiver<P>) {}
  get id(): number { return this.raw.id; }
  async next(options?: RequestOptions | number): Promise<InboundFrame | null> {
    const frame = await this.raw.next(options);
    if (!frame) return null;
    try { return codecs.get(this.raw)!.decode(frame); }
    catch { throw new ClientError("invalid_data", "unknown"); }
  }
  close(): void { this.raw.close(); }
  async *[Symbol.asyncIterator](): AsyncIterator<InboundFrame> {
    try { for (;;) { const frame = await this.next(); if (!frame) return; yield frame; } }
    finally { this.close(); }
  }
}

/** Message subscription with native sends and optional ownership-bearing split. */
export class Stream<P extends Protocol> implements AsyncIterable<InboundFrame> {
  private parts?: { sender: StreamSender<P>; receiver: StreamReceiver<P> };
  constructor(raw: RawStream<P>) {
    const parts = raw.split();
    this.parts = { sender: new StreamSender(parts.sender), receiver: new StreamReceiver(parts.receiver) };
  }
  get id(): number { return this.owned().sender.id; }
  send(message: OutboundMessage, options?: RequestOptions): Promise<void> { return this.owned().sender.send(message, options); }
  next(options?: RequestOptions | number): Promise<InboundFrame | null> { return this.owned().receiver.next(options); }
  close(): void { if (this.parts) { this.parts.receiver.close(); this.parts.sender.close(); } }
  split(): { sender: StreamSender<P>; receiver: StreamReceiver<P> } { const parts = this.owned(); this.parts = undefined; return parts; }
  private owned() { if (!this.parts) throw new ClientError("split"); return this.parts; }
  async *[Symbol.asyncIterator](): AsyncIterator<InboundFrame> {
    try { for (;;) { const frame = await this.next(); if (!frame) return; yield frame; } }
    finally { this.close(); }
  }
}
